<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.5"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>scran.js: Single cell RNA-seq analysis in Javascript</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">scran.js
   </div>
   <div id="projectbrief">Compiling scran&#39;s C++ library to WebAssembly</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.5 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
</div><!-- top -->
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div><div class="header">
  <div class="headertitle"><div class="title">Single cell RNA-seq analysis in Javascript </div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p ><a class="anchor" id="md__github_workspace_README"></a> </p><h1>Overview</h1>
<p >This repository contains <b>scran.js</b>, a Javascript library for single-cell RNA-seq (scRNA-seq) analysis in the browser. The various calculations are performed directly by the client, allowing us to take advantage of the ubiquity of the browser as a standalone analysis platform. Users can then directly analyze their data without needing to manage any dependencies or pay for access to a backend. <b>scran.js</b> is heavily inspired by the <a href="https://bioconductor.org/packages/scran"><b>scran</b> R package</a> and contains most of its related methods. Indeed, much of the implementation in this repository is taken directly from <b>scran</b> and its related R packages.</p>
<h1>Analysis overview</h1>
<p ><b>scran.js</b> implements key steps in a typical scRNA-seq analysis:</p>
<ul>
<li>Quality control (QC) to remove low-quality cells. This is done based on detection of outliers on QC metrics like the number of detected genes.</li>
<li>Normalization and log-transformation, to remove biases and mitigate the mean-variance trend. We use scaling normalization with size factors defined from the library size for each cell.</li>
<li>Feature selection to identify highly variable genes. This is based on residuals from a trend fitted to the means and variances of the log-normalized data for each gene.</li>
<li>Principal components analysis (PCA) on the highly variable genes, to compress and denoise the data. We use an approximate method to quickly obtain the top few PCs.</li>
<li>Clustering on the cells using the top PCs. This can either use k-means or community detection on a shared nearest neighbor graph.</li>
<li>Dimensionality reduction with t-stochastic neighbor embedding (t-SNE), again using the top PCs.</li>
<li>Marker detection using a variety of effect sizes such as Cohen's d and the area under the curve (AUC).</li>
<li>Cell type annotation with a port of the <a href="https://bioconductor.org/packages/SingleR"><b>SingleR</b></a> algorithm.</li>
<li>Batch correction via the mutual nearest neighbors method.</li>
</ul>
<p >The theory behind these methods is described in more detail in the <a href="https://bioconductor.org/books/release/OSCA/"><b>Orchestrating Single Cell Analysis with Bioconductor</b></a> book. All steps are implemented in C++ and compiled to WebAssembly for near-native performance - see the developer notes for details.</p>
<h1>Quick start</h1>
<p ><b>scran.js</b> is available as an <a href="https://www.npmjs.com/package/scran.js">npm package</a>, so installation can be performed via the usual procedure:</p>
<div class="fragment"><div class="line">npm install scran.js</div>
</div><!-- fragment --><p >Then you can import and initialize the library using the standard ES6 notation. Before any actual analysis steps are performed, the <code>initialize()</code> function must be run and its promise resolved; we suggest using a top-level <code>await</code> for convenience.</p>
<div class="fragment"><div class="line">import * as scran from &quot;scran.js&quot;;</div>
<div class="line">await scran.initialize({ numberOfThreads: 4 }); // for old Node versions, set localFile: true</div>
</div><!-- fragment --><p >After that, you can run the remaining steps synchronously - for example, using the Node.js API:</p>
<div class="fragment"><div class="line">// Reading in the count matrix.</div>
<div class="line">import * as fs from &quot;fs&quot;;</div>
<div class="line">let buffer = fs.readFileSync(&quot;matrix.mtx.gz&quot;);</div>
<div class="line">let mat = scran.initializeSparseMatrixFromMatrixMarketBuffer(buffer);</div>
</div><!-- fragment --><h1>Basic analyses</h1>
<p >The code chunk below implements a basic analysis, starting from count matrix loading and terminating at the markers for each cluster. This uses the Node.js API to read in one of our example <a href="https://github.com/jkanche/random-test-files">Matrix Market files</a>, but the same approach can be used on the buffer created from a <code>File</code> input in the browser.</p>
<div class="fragment"><div class="line">import * as scran from &quot;scran.js&quot;;</div>
<div class="line">await scran.initialize({ numberOfThreads: 4 });</div>
<div class="line"> </div>
<div class="line">// Reading in the count matrix.</div>
<div class="line">import * as fs from &quot;fs&quot;;</div>
<div class="line">let buffer = fs.readFileSync(&quot;matrix.mtx.gz&quot;);</div>
<div class="line">let input = scran.initializeSparseMatrixFromMatrixMarketBuffer(buffer);</div>
<div class="line">let mat = input.matrix;</div>
<div class="line"> </div>
<div class="line">// Performing QC.</div>
<div class="line">let qc_metrics = scran.computePerCellQCMetrics(mat, [ /* specify mito subset here */ ]);</div>
<div class="line">let qc_thresholds = scran.computePerCellQCFilters(qc_metrics);</div>
<div class="line">let filtered = scran.filterCells(mat, qc_thresholds.discardOverall());</div>
<div class="line"> </div>
<div class="line">// Log-normalizing.</div>
<div class="line">let normalized = scran.logNormCounts(filtered);</div>
<div class="line"> </div>
<div class="line">// Modelling per-gene variance and selecting top HVGs. </div>
<div class="line">let varmodel = scran.modelGeneVar(normalized);</div>
<div class="line">let hvgs = scran.chooseHVGs(varmodel, { number: 4000 });</div>
<div class="line"> </div>
<div class="line">// Performing the PCA.</div>
<div class="line">let pcs = scran.runPCA(normalized, { features: hvgs });</div>
<div class="line"> </div>
<div class="line">// Building the neighbor search index on the PCs.</div>
<div class="line">let index = scran.buildNeighborSearchIndex(pcs);</div>
<div class="line"> </div>
<div class="line">// Performing the clustering. </div>
<div class="line">let cluster_graph = scran.buildSNNGraph(index, { neighbors: 10 });</div>
<div class="line">let clustering = scran.clusterSNNGraph(cluster_graph);</div>
<div class="line"> </div>
<div class="line">// Performing the t-SNE and UMAP.</div>
<div class="line">let tsne_status = scran.initializeTSNE(index);</div>
<div class="line">let tsne_res = scran.runTSNE(tsne_status);</div>
<div class="line"> </div>
<div class="line">let umap_status = scran.initializeUMAP(index);</div>
<div class="line">let umap_res = scran.runUMAP(umap_status);</div>
<div class="line"> </div>
<div class="line">// Detecting some markers.</div>
<div class="line">let markers = scran.scoreMarkers(normalized, clustering.membership());</div>
<div class="line"> </div>
<div class="line">scran.terminate();</div>
</div><!-- fragment --><h1>More documentation</h1>
<p >Reference documentation for the Javascript API is available <a href="https://jkanche.github.io/scran.js">here</a>.</p>
<p >Specific topics are discussed in more detail below:</p>
<ul>
<li>Terminate upon completion</li>
<li>Manually free memory</li>
<li>Use k-means clustering</li>
<li>Annotate cell types against references</li>
<li>Handling the permuted row order</li>
</ul>
<p >Developer notes are also available here.</p>
<h1>Links</h1>
<p >Check out <a href="https://github.com/jkanche/kana">kana</a> to see how <b>scran.js</b> is used in an interactive scRNA-seq analysis application.</p>
<p >The <a href="https://github.com/LTLA/scran.chan"><b>scran.chan</b></a> R package and <a href="https://github.com/LTLA/scran-cli"><b>scran</b></a> executable are based on the same C++ libraries and allow the same analysis to be performed in different environments. </p>
</div></div><!-- PageDoc -->
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.5
</small></address>
</body>
</html>
